amivapi/auth/apikeys.py

```diff
@@ -3,38 +3,107 @@
 # license: AGPLv3, see LICENSE for details. In addition we strongly encourage
 #          you to buy us beer if we meet and you like the software.
 
-"""Hook to allow authorization with apikeys."""
+"""Authorization with apikeys.
+
+Provides the apikey resource and hooks to handle authorization with a key.
+
+API keys should only be created or modified by admins.
+"""
+from os import urandom
+from base64 import b64encode
+from datetime import datetime as dt
 
 from flask import abort, current_app, g
 
-from ruamel import yaml
+from amivapi.utils import register_domain
+
+from .auth import AmivTokenAuth
 
 
 def authorize_apikeys(resource):
-    """Check if user is an apikey, and if it is, do authorization."""
-    permissions = [
-        data['permissions'] for data in current_app.config['APIKEYS'].values()
-        if data['token'] == g.get('current_token')]
-    if permissions:
-        # We can just take first (and only) element since token is unique
-        permission = permissions[0].get(resource)
+    """Check if user is an apikey, and if it is, do authorization.
+
+    Also update 'updated' timestamp everytime a key is accessed
+    """
+    apikeys = current_app.data.driver.db['apikeys']
+    apikey = apikeys.find_one({'token': g.get('current_token')},
+                              {'permissions': 1})
+
+    if apikey:
+        # Get permission for resource if they exist
+        permission = apikey['permissions'].get(resource)
+
+        # Update timestamp (remove microseconds to match mongo precision)
+        new_time = dt.utcnow().replace(microsecond=0)
+        apikeys.update_one({'_id': apikey['_id']},
+                           {'$set': {'_updated': new_time}})
 
         if permission == 'read':
             g.resource_admin_readonly = True
         elif permission == 'readwrite':
             g.resource_admin = True
         elif g.get('auth_required'):
-            abort(403, "The APIKEY does not grant the required permissions.")
+            abort(403, "The API key exists but does not grant the required "
+                       "permissions.")
+
+
+class ApikeyAuth(AmivTokenAuth):
+    """Do not let (non-admin) users view API keys at all.
+
+    We don't want to filter the output , we want to abort completely."""
+    def create_user_lookup_filter(self, user):
+        abort(403)
```

Comment:
I don't quite understand your question. What is weird about it? This is unchanged from the previous implementation of api keys. 

And yes, the mechanism is practically identical to groups, i.e. Api keys grant permissions in the same ways groups do. Nevertheless groups are only for real users and api keys exists to give a set of permissions to applications. 

What part exactly do you think should be done in groups? And how?